### PR TITLE
feat(lib/parachain): Implement request and response message for /req_collation/1 protocol

### LINKED ISSUE #3339: Implement request and response message for `/req_collation/1` protocol
## Issue summary
- Should be done after https://github.com/ChainSafe/gossamer/pull/3334/files is merged
- Request struct must implement `network.Message` interface
- Response struct must implement `network.ResponseMessage`

## Other information and links
Links to the spec

- https://spec.polkadot.network/chapter-anv#net-msg-collation-fetching-request
- https://github.com/paritytech/polkadot/blob/47c8c36d6e3f90da77ebccb8b672e4fb4f33c84c/node/network/protocol/src/request_response/outgoing.rs#L34
- https://spec.polkadot.network/chapter-anv#net-msg-collation-fetching-response

### PR DESCRIPTION
**Pull request by axaysagathiya**
## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues

Closes #3339 
## Primary Reviewer

@kanishkatn @edwardmack

### CODE REVIEW

File: lib/parachain/collation_fetching.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+type CollationFetchingRequest struct {
+	// Relay parent we want a collation for
+	RelayParent common.Hash
+
+	// Parachain id of the collation
+	ParaID ParaID
+}
+
+// Encode returns the SCALE encoding of the CollationFetchingRequest
+func (c *CollationFetchingRequest) Encode() ([]byte, error) {
```
Comment (jimjbrettj) [inline lib/parachain/collation_fetching.go:21]: Do we need to wrap marshal with an encode func? Typically I think we have avoided this unless the encoding in particularly complex/not scale

File: lib/parachain/collation_fetching.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+type CollationFetchingResponse scale.VaryingDataType
+
+// Collation represents a requested collation to be delivered
+type Collation struct {
+	CandidateReceipt CandidateReceipt `scale:"1"`
+	PoV              PoV              `scale:"2"`
+}
+
+// PoV represents a Proof-of-Validity
+type PoV struct {
```
Comment (jimjbrettj) [inline lib/parachain/collation_fetching.go:35]: This isnt a vdt is it? Why struct with just one element?

File: lib/parachain/collation_fetching.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// Value returns the value from the underlying varying data type
+func (c *CollationFetchingResponse) Value() (val scale.VaryingDataTypeValue, err error) {
+	vdt := scale.VaryingDataType(*c)
+	return vdt.Value()
+}
+
+// Encode returns the SCALE encoding of the CollationFetchingResponse
+func (c *CollationFetchingResponse) Encode() ([]byte, error) {
```
Comment (jimjbrettj) [inline lib/parachain/collation_fetching.go:68]: Again, not sure if these wrappers are needed

**Review by jimjbrettj — COMMENTED** [review]
I think we do not need wrappers around scale here, but otherwise looks good to me

File: lib/parachain/collation_fetching.go (hunk 2 above)
Comment (axaysagathiya) [reply lib/parachain/collation_fetching.go:35]: I think I can change it to `type PoV []byte`

File: lib/parachain/collation_fetching.go (hunk 1 above)
Comment (axaysagathiya) [reply lib/parachain/collation_fetching.go:21]: I have implemented the 'network.Message' interface as 'CollationFetchingRequest' will be passed into [this](https://github.com/ChainSafe/gossamer/pull/3334/files#diff-5794e04e36068d98b7c6d2280c7e0b41fecca0b90bd6dd4576cd62b951ab587aR32) function as 'req'.

File: lib/parachain/collation_fetching.go (hunk 3 above)
Comment (axaysagathiya) [reply lib/parachain/collation_fetching.go:68]: I have implemented the '[network.ResponseMessage](https://github.com/ChainSafe/gossamer/pull/3334/files#diff-5794e04e36068d98b7c6d2280c7e0b41fecca0b90bd6dd4576cd62b951ab587aR88)' interface as 'CollationFetchingResponse' will be passed into [this](https://github.com/ChainSafe/gossamer/pull/3334/files#diff-5794e04e36068d98b7c6d2280c7e0b41fecca0b90bd6dd4576cd62b951ab587aR32) function as 'res'.

File: lib/parachain/collation_fetching.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+type PoV []byte
+
+// Index returns the index of varying data type
+func (Collation) Index() uint {
+	return 0
+}
+
+// NewCollationFetchingResponse returns a new collation fetching response varying data type
+func NewCollationFetchingResponse() CollationFetchingResponse {
+	vdt := scale.MustNewVaryingDataType(Collation{})
```
Comment (kanishkatn) [inline lib/parachain/collation_fetching.go:45]: Why not return an error instead?

File: lib/parachain/collation_fetching.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+// Collation represents a requested collation to be delivered
+type Collation struct {
+	CandidateReceipt CandidateReceipt `scale:"1"`
+	PoV              PoV              `scale:"2"`
+}
+
+// PoV represents a Proof-of-Validity block (PoV block) or a parachain block.
+// It contains the necessary data for the parachain specific state transition logic.
+type PoV []byte
```
Comment (EclesioMeloJunior) [inline lib/parachain/collation_fetching.go:36]: do we need this as a type? why not keep the `Collation.PoV` as just `[]byte`

**Review by EclesioMeloJunior — APPROVED** [review]
LGMT! Just a question around the `PoV` type

File: lib/parachain/collation_fetching.go (hunk 5 above)
Comment (axaysagathiya) [reply lib/parachain/collation_fetching.go]: Not really, changed it.

**Review by kanishkatn — APPROVED** [review]

File: lib/parachain/collation_fetching_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+import (
+	"testing"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEncodeCollationFetchingRequest(t *testing.T) {
+	collationFetchingRequest := CollationFetchingRequest{
+		RelayParent: getDummyHash(6),
```
Comment (kishansagathiya) [inline lib/parachain/collation_fetching_test.go:12]: let's use a realistic hash value.

File: lib/parachain/collation_fetching_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	require.NoError(t, err)
+
+	expextedEncode := common.MustHexToBytes("0x060606060606060606060606060606060606060606060606060606060606060605000000")
+	require.Equal(t, expextedEncode, actualEncode)
+}
+
+func TestCollationFetchingResponse(t *testing.T) {
+	t.Parallel()
+
+	hash6 := getDummyHash(6)
```
Comment (kishansagathiya) [inline lib/parachain/collation_fetching_test.go:26]: realistic hash

**Review by kishansagathiya — APPROVED** [review]

**Pull request merged**